cartoframes/data/services/geocode.py

```diff
@@ -0,0 +1,484 @@
+            the_geom_webmercator = CASE
+              WHEN _g.the_geom IS NULL THEN NULL
+              ELSE ST_Transform(_g.the_geom, 3857)
+            END,
+            {metadata_assignment}
+            {hash_column} = {hash_expression}
+        FROM (SELECT * FROM {geocode_expression}) _g
+        WHERE _g.cartodb_id = {table}.cartodb_id
+    """.format(
+        table=table,
+        hash_column=HASH_COLUMN,
+        hash_expression=hash_expression,
+        geocode_expression=geocode_expression,
+        metadata_assignment=metadata_assignment
+    )
+
+
+def _hash_as_big_int(text):
+    # Calculate a positive bigint hash, hence the 63-bit mask.
+    return int(hashlib.sha1(text.encode()).hexdigest(), 16) & ((2**63)-1)
+
+
+def _generate_temp_table_name(base=None):
+    return (base or 'table') + '_' + uuid.uuid4().hex[:10]
+
+
+def _set_pre_summary_info(summary, output):
+    logging.debug(summary)
+    output['total_rows'] = sum(summary.values())
+    # TODO: either report new (ng+nn) and changed (cg+cn) records, or we could
+    # reduce the states by merging ng+cg, nn+nn
+    output['required_quota'] = sum([summary[s] for s in ['ng', 'nn', 'cg', 'cn']])
+    output['previously_geocoded'] = summary.get('pg', 0)
+    output['previously_failed'] = summary.get('pn', 0)
+    output['records_with_geometry'] = sum([summary[s] for s in ['ng', 'cg', 'pg']])
+    # output['records_without_geometry'] = sum([summary[s] for s in ['nn', 'cn', 'pn']])
+
+
+def _set_post_summary_info(summary, result, output):
+    if result and result.get('total_rows', 0) == 1:
+        null_geom_count = result.get('rows')[0].get('count')
+        geom_count = output['total_rows'] - null_geom_count
+        output['final_records_with_geometry'] = geom_count
+        # output['final_records_without_geometry'] = null_geom_count
+        output['geocoded_increment'] = output['final_records_with_geometry'] - output['records_with_geometry']
+        output['successfully_geocoded'] = output['geocoded_increment'] + sum([summary[s] for s in ['ng', 'cg']])
+        output['failed_geocodings'] = output['required_quota'] - output['successfully_geocoded']
+
+
+def _dup_dataset(dataset):
+    # When uploading datasets to temporary tables we don't want to leave the dataset
+    # with a `table_name` attribute (of the temporary table, that will be deleted),
+    # so we'll use duplicates of the datasets for temporary uploads`
+    if hasattr(dataset, 'query'):
+        return Dataset(dataset.query)
+    elif dataset.table_name:
+        return Dataset(dataset.table_name)
+    return Dataset(dataset.dataframe)
+
+
+class Geocode(object):
+    """Geocode using CARTO data services.
+    This requires a CARTO account; master API Key credentials must be provided
+    (throuch explicit argument in contructor or via the default credentials)
+    to access the service. Use of these methods will incurn in geocoding
+    credit consumption for the privided accout.
+
+    Examples:
```

Comment:
it would be useful to add the credentials part (how to do it) in every example that needs it:

```
from cartoframes.auth import set_default_credentials

set_default_credentials('your_user_name', 'your api key')
```